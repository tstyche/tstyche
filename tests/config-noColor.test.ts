import { afterEach, describe, expect, test } from "@jest/globals";
import ansiEscapesSerializer from "jest-serializer-ansi-escapes";
import { clearFixture, writeFixture } from "./__utils__/fixtureFactory.js";
import { normalizeOutput } from "./__utils__/normalizeOutput.js";
import { spawnTyche } from "./__utils__/spawnTyche.js";

expect.addSnapshotSerializer(ansiEscapesSerializer);

const isStringTestText = `import { expect, test } from "tstyche";
test("is string?", () => {
  expect<string>().type.toBeString();
});
`;

const tsconfig = {
  extends: "../tsconfig.json",
  include: ["./"],
};

const fixture = "config-noColor";

afterEach(async () => {
  await clearFixture(fixture);
});

describe("'TSTYCHE_NO_COLOR' environment variable", () => {
  test("has default value", async () => {
    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["--showConfig"], {
      env: { ["TSTYCHE_NO_COLOR"]: undefined },
    });

    expect(JSON.parse(stdout)).toHaveProperty("noColor");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when truthy, colors are disabled", async () => {
    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, [], {
      env: { ["TSTYCHE_NO_COLOR"]: "true" },
    });

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when falsy, colors are enabled", async () => {
    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, [], {
      env: { ["TSTYCHE_NO_COLOR"]: "" },
    });

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when 'NO_COLOR' is truthy, colors are disabled", async () => {
    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, [], {
      env: {
        ["NO_COLOR"]: "true",
        ["TSTYCHE_NO_COLOR"]: undefined,
      },
    });

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when 'NO_COLOR' is falsy, colors are enabled", async () => {
    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, [], {
      env: {
        ["NO_COLOR"]: "",
        ["TSTYCHE_NO_COLOR"]: undefined,
      },
    });

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("overrides 'NO_COLOR' and enables colors", async () => {
    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, [], {
      env: {
        ["NO_COLOR"]: "true",
        ["TSTYCHE_NO_COLOR"]: "",
      },
    });

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("overrides 'NO_COLOR' and disables colors", async () => {
    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, [], {
      env: {
        ["NO_COLOR"]: "",
        ["TSTYCHE_NO_COLOR"]: "true",
      },
    });

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });
});
