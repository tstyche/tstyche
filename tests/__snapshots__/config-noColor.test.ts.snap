// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'TSTYCHE_NO_COLOR' environment variable overrides 'NO_COLOR' and disables colors: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  + is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      1 passed, 1 total
Assertions: 1 passed, 1 total
Duration:   <<timestamp>>

Ran all test files.
"
`;

exports[`'TSTYCHE_NO_COLOR' environment variable overrides 'NO_COLOR' and enables colors: stdout 1`] = `
"<blue>uses</> TypeScript 5.3.3<gray> with ./tsconfig.json</>

<green>pass</> <gray>./__typetests__/</>dummy.test.ts
  <green>+</> <gray>is string?</>

Targets:    <green>1 passed</>, 1 total
Test files: <green>1 passed</>, 1 total
Tests:      <green>1 passed</>, 1 total
Assertions: <green>1 passed</>, 1 total
Duration:   <<timestamp>>

<gray>Ran </><gray>all test files.</>
"
`;

exports[`'TSTYCHE_NO_COLOR' environment variable when 'NO_COLOR' is falsy, colors are enabled: stdout 1`] = `
"<blue>uses</> TypeScript 5.3.3<gray> with ./tsconfig.json</>

<green>pass</> <gray>./__typetests__/</>dummy.test.ts
  <green>+</> <gray>is string?</>

Targets:    <green>1 passed</>, 1 total
Test files: <green>1 passed</>, 1 total
Tests:      <green>1 passed</>, 1 total
Assertions: <green>1 passed</>, 1 total
Duration:   <<timestamp>>

<gray>Ran </><gray>all test files.</>
"
`;

exports[`'TSTYCHE_NO_COLOR' environment variable when 'NO_COLOR' is truthy, colors are disabled: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  + is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      1 passed, 1 total
Assertions: 1 passed, 1 total
Duration:   <<timestamp>>

Ran all test files.
"
`;

exports[`'TSTYCHE_NO_COLOR' environment variable when falsy, colors are enabled: stdout 1`] = `
"<blue>uses</> TypeScript 5.3.3<gray> with ./tsconfig.json</>

<green>pass</> <gray>./__typetests__/</>dummy.test.ts
  <green>+</> <gray>is string?</>

Targets:    <green>1 passed</>, 1 total
Test files: <green>1 passed</>, 1 total
Tests:      <green>1 passed</>, 1 total
Assertions: <green>1 passed</>, 1 total
Duration:   <<timestamp>>

<gray>Ran </><gray>all test files.</>
"
`;

exports[`'TSTYCHE_NO_COLOR' environment variable when truthy, colors are disabled: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  + is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      1 passed, 1 total
Assertions: 1 passed, 1 total
Duration:   <<timestamp>>

Ran all test files.
"
`;
